Reuse running adb instead of killing existing one. Fixes #464

bkase opened the pull request:
This patch reuses a running `adb` process if one exists. (fixes #464)

Tested on Arch Linux, Ubuntu, Windows, and OSX. To test, I ran the simulator add-on with `adb logcat` running in another terminal. And again without `adb logcat`. When the module unloads and `adb` was running on startup, `adb` will be killed. When the module unloads and `adb` was not running on startup, `adb` won't be killed.

bkase committed:
Reuse running adb instead of killing existing one

mykmelez commented on addon/lib/adb.js:
```diff
@@ -21,6 +21,30 @@ let Cc = components.classes;
+  });
+
+  // This variable needs to be named something other than require or else
+  // the addon-sdk loader gets confused
```
Can you elaborate about how it gets confused? It'd be useful to have additional information here about precisely what the issue is.

mykmelez commented on addon/lib/simulator.js:
```diff
@@ -69,7 +69,9 @@ let simulator = module.exports = {
     this.kill();
-    ADB.kill(Runtime.OS == "WINNT" ? true : false /* sync */);
+    if (ADB.didRunInitially) {
+      ADB.kill(Runtime.OS == "WINNT" ? true : false /* sync */);
```
Calling _ADB.kill_ should set _didRunInitially_ back to `false`.

bkase committed:
Accidentally kept typo in ps command

It was there in the first place to test the behavior if `ps` does not
exist somewhere in PATH

bkase committed:
Elaborated on addon-sdk confusion over `require`

bkase committed:
Calling ADB.kill resets didRunInitially

mykmelez commented:
On my Mac, opening the Dashboard with these changes applied prints the following messages:

> info: r2d2b2g: adb: Error: a task list executable not found on filesystem
> info: r2d2b2g: adb: Didn't find ADB process running, restarting

Whereas the ADB module should find and use the _ps_ executable at `/bin/ps`.

Moreover, "push to device" fails. Pressing the _Push_ button for an app prints:

> info: r2d2b2g: Simulator.onMessage pushAppToDevice
> info: r2d2b2g: Simulator.pushAppToDevice: /Users/myk/Mozilla/GeoLocTest/manifest.webapp
> info: r2d2b2g: adb: forwardPort 6000
> info: r2d2b2g: adb: runCommand host:forward:tcp:6000;tcp:6000

But the device never asks me to confirm an incoming debugging connection, and the app never gets pushed.

bkase commented:
I accidentally kept a typo in the "ps" file when I sent out the pull-request, that should be fixed now.

I'm looking into the push to device problem

mykmelez commented:
I just tested on Windows 7, and I see the "push to device" problem there too.

bkase committed:
Fixed push-to-device problem

There was a `self` reference when there should have been a `this`

bkase commented:
I never tested the push-to-device functionality, I assumed that if the dashboard said the device was connected, everything would work. Sorry about that.
It definitely works on Mac now, I'll check on the other OSs quickly.

EDIT: Works for me on OSX, Ubuntu, and Arch Linux, but I don't have adb drivers working for Windows yet

mykmelez commented:
Looks good, thanks for the fixes!
